### PR TITLE
coordinator dynamic config POST to update only explicitly specified fields

### PR DESCRIPTION
**Pull request by himanshug**
...instead of resetting everything else to zeros

currently if user makes a mistake and does not include some fields then they are set to zeros causing major problems afterwards.

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 1]
```diff
@@ -45,20 +46,60 @@
...
-      @JsonProperty("killAllDataSources") boolean killAllDataSources
+      @JsonProperty("killAllDataSources") Boolean killAllDataSources
+  )
+  {
+    CoordinatorDynamicConfig current = configManager.watch(
+        CoordinatorDynamicConfig.CONFIG_KEY,
+        CoordinatorDynamicConfig.class
+    ).get();
+    if (current == null) {
+      current = new Builder().build();
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: This line can be moved to ```configManager.watch()``` as a parameter for default value.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 2]
```diff
@@ -45,20 +46,60 @@
...
+    }
+
+    this.millisToWaitBeforeDeleting = millisToWaitBeforeDeleting == null ? current.getMillisToWaitBeforeDeleting() : millisToWaitBeforeDeleting;
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: This line is too long. Would you change this according to our format?

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 3]
```diff
@@ -45,20 +46,60 @@
...
+    this.mergeBytesLimit = mergeBytesLimit == null ? current.getMergeBytesLimit() : mergeBytesLimit;
+    this.mergeSegmentsLimit = mergeSegmentsLimit == null ? current.getMergeSegmentsLimit() : mergeSegmentsLimit;
+    this.maxSegmentsToMove = maxSegmentsToMove == null ? current.getMaxSegmentsToMove() : maxSegmentsToMove;
+    this.replicantLifetime = replicantLifetime == null ? current.getReplicantLifetime() : replicantLifetime;
+    this.replicationThrottleLimit = replicationThrottleLimit == null ? current.getReplicationThrottleLimit() : replicationThrottleLimit;
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Format here too.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 4]
```diff
@@ -45,20 +46,60 @@
...
+    this.balancerComputeThreads = Math.max(balancerComputeThreads == null ? current.getBalancerComputeThreads() : balancerComputeThreads, 1);
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Format here too.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 5]
```diff
@@ -45,20 +46,60 @@
...
+    this.emitBalancingStats = emitBalancingStats == null ? current.emitBalancingStats() : emitBalancingStats;
+
+
+    this.killAllDataSources = killAllDataSources == null ? current.isKillAllDataSources() : killAllDataSources;
+    this.killDataSourceWhitelist = killDataSourceWhitelist == null ? current.getKillDataSourceWhitelist() : parseKillDataSourceWhitelist(killDataSourceWhitelist);
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Format here too.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 6]
```diff
@@ -45,20 +46,60 @@
...
+
+    if (this.killAllDataSources && !this.killDataSourceWhitelist.isEmpty()) {
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Seems that DruidCoordinatorSegmentKiller doesn't have to check the same condition [here](https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentKiller.java#L88). Would you remove it?

**Comment by fjy** [comment]
👍  after formatting fixes

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 1 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: not really, watch() returns a AtomicReference "watcher" which could return null at some point but a valid value later if config is modified in the database by anything.
we could add something configManager.get(T default) to get current value at some point but watch().get() is decent workaround and is used in most places.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 2 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: sure

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 3 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: sure

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 4 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: sure

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 6 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: yeah the check could be removed from DruidCoordinatorSegmentKiller but given that hard segment deletion is a very sensitive area I'm inclined to keep the check there ... it does no harm.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 5 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: sure

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 1 above)
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Sorry for my late reply. I meant, the below two codes are equivalent and the later one would be better to read.

```
CoordinatorDynamicConfig current = configManager.watch(
        CoordinatorDynamicConfig.CONFIG_KEY,
        CoordinatorDynamicConfig.class
    ).get();
    if (current == null) {
      current = new Builder().build();
    }
```

```
CoordinatorDynamicConfig current = configManager.watch(
        CoordinatorDynamicConfig.CONFIG_KEY,
        CoordinatorDynamicConfig.class,
        new Builder().build()
    ).get();
```

**Comment by jihoonson** [comment]
The patch looks good to me. I left a trivial comment, but I'm fine with as it is.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 1 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: @jihoonson I did not want to build the default one unless needed, with this code default config will always be created.

Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Ok.

**Pull request merged**